lib/auth/join_bound_keypair.go

```diff
@@ -0,0 +1,489 @@
+	hasIncomingBotInstance := req.JoinRequest.BotInstanceID != ""
+	hasJoinsRemaining := status.JoinCount < spec.Joining.TotalJoins
+
+	// if set, the bound bot instance will be updated in the backend
+	expectNewBotInstance := false
+
+	// the bound public key; may change during initial join or rotation. used to
+	// inform the returned public key value.
+	boundPublicKey := status.BoundPublicKey
+
+	// Mutators to use during the token resource status patch at the end.
+	var mutators []boundKeypairStatusMutator
+
+	switch {
+	case !hasBoundPublicKey && !hasIncomingBotInstance:
+		// Normal initial join attempt. No bound key, and no incoming bot
+		// instance. Consumes a rejoin.
+		if spec.Onboarding.InitialJoinSecret != "" {
+			return nil, "", trace.NotImplemented("initial joining secrets are not yet supported")
+		}
+
+		if spec.Onboarding.InitialPublicKey == "" {
+			return nil, "", trace.BadParameter("an initial public key is required")
+		}
+
+		if !spec.Joining.Unlimited && !hasJoinsRemaining {
+			return nil, "", trace.AccessDenied("no joins remaining")
+		}
+
+		if err := a.issueBoundKeypairChallenge(
+			ctx,
+			spec.Onboarding.InitialPublicKey,
+			challengeResponse,
+		); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		// Now that we've confirmed the key, we can consider it bound.
+		mutators = append(
+			mutators,
+			mutateStatusBoundPublicKey(spec.Onboarding.InitialPublicKey, ""),
+			mutateStatusConsumeJoin(spec.Joining.Unlimited, status.JoinCount, spec.Joining.TotalJoins),
+		)
+
+		expectNewBotInstance = true
+		boundPublicKey = spec.Onboarding.InitialPublicKey
+	case !hasBoundPublicKey && hasIncomingBotInstance:
+		// Not allowed, at least at the moment. This would imply e.g. trying to
+		// change auth methods.
+		return nil, "", trace.BadParameter("cannot perform first bound keypair join with existing credentials")
+	case hasBoundPublicKey && !hasBoundBotInstance:
+		// TODO: Bad backend state, or maybe an incomplete previous join
+		// attempt. This shouldn't be a possible state, but we should handle it
+		// sanely anyway.
+		return nil, "", trace.BadParameter("bad backend state, please recreate the join token")
+	case hasBoundPublicKey && hasBoundBotInstance && hasIncomingBotInstance:
+		// Standard rejoin case, does not consume a rejoin.
+		if status.BoundBotInstanceID != req.JoinRequest.BotInstanceID {
+			return nil, "", trace.AccessDenied("bot instance mismatch")
+		}
+
+		if err := a.issueBoundKeypairChallenge(
+			ctx,
+			spec.Onboarding.InitialPublicKey,
+			challengeResponse,
+		); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		// Nothing else to do, no key change
+	case hasBoundPublicKey && hasBoundBotInstance && !hasIncomingBotInstance:
+		// Hard rejoin case, the client identity expired and a new bot instance
+		// is required. Consumes a rejoin.
+		if !spec.Joining.Unlimited && !hasJoinsRemaining {
+			return nil, "", trace.AccessDenied("no rejoins remaining")
+		}
```

Comment:
Should this check actually be inside the function passed to `PatchToken` in case the join counter has been modified concurrently?

Reply:
The decision is made here, but the mutator function (`mutateStatusConsumeJoin`) checks that:
- the join (now recovery) counter has not changed since the decision was made
- the join (now recovery) limit is at least the value it was when the decision was made

The mutator functions are run inside `PatchToken`, so those checks will always need to pass for the patch to succeed.